common/lib/xmodule/xmodule/library_content_module.py

```diff
@@ -371,13 +389,159 @@ def get_child_descriptors(self):
+        so we only remove blocks that are no longer valid here.
+        """
+        selected = set(tuple(k) for k in selected)  # set of (block_type, block_id) tuples assigned to this student
+
+        # Determine which of our children we will show:
+        valid_block_keys = set([(c.block_type, c.block_id) for c in children])
+        # Remove any selected blocks that are no longer valid:
+        invalid_block_keys = (selected - valid_block_keys)
+        if invalid_block_keys:
+            selected -= invalid_block_keys
+
+        # `overlimit_block_keys`:
+        # External service determines number of children to show,
+        # so no need to *drop* any blocks based on value of `max_count` argument.
+
+        # `added_block_keys`:
+        # External service determines number of children to show,
+        # so no need to *add* any blocks based on value of `max_count` argument.
+
+        return {
+            'selected': selected,
+            'invalid': invalid_block_keys,
+            # External service determines number of children to show,
+            # so `max_count` setting has no effect:
+            # We never *drop* previously selected children because of it,
+            # so the set of dropped children is empty by definition:
+            'overlimit': set(),
+            # External service determines number of children to show,
+            # so `max_count` setting has no effect:
+            # We never *add* any children because of it,
+            # so the set of added children is empty by definition:
+            'added': set(),
+        }
+
+    def selected_children(self):
+        """
+        Returns a set() of block_ids indicating which of the possible children
+        have been selected to display to the current user.
+        """
+        if hasattr(self, "_selected_set"):
+            # Already done:
+            return self._selected_set  # pylint: disable=access-member-before-definition
+
+        # Determine children to display based on information about pending reviews from external service.
+        # This needs to happen here (at instance level) because we need access to course and block-specific data,
+        # which is not available at the class level.
+        selected = self.get_selections_current_user(self.children)
+        block_keys = self.make_selection(selected, self.children, self.max_count, "adaptive")  # pylint: disable=no-member
+
+        # Publish events for analytics purposes:
+        self.publish_events(block_keys)
+
+        # Store selected children
+        selected = block_keys['selected']
+        self.store_selected(selected)
+
+        return selected
+
+    def get_selections_current_user(self, children):
+        """
+        Check with external service whether any children of this block
+        are scheduled for review by the current user, and return them.
+        """
+        pending_reviews = self.get_pending_reviews(self.anonymous_user_id)
```

Comment:
Got `ValueError: No JSON object could be decoded` when viewing the ACB unit in the LMS as a student user.  I tried it with 2 different new student accounts on my devstack, and got the same results.

No error is shown for my staff user though.

Full trace below:

```
2016-12-26 07:34:07,620 ERROR 3795 [django.request] base.py:256 - Internal Server Error: /courses/course-v1:OpenCraft+ACB1+OC-2150/courseware/519f9f58a85c4d2aba6182162b1c79e8/e94c284e0a104b77987a3a872367e9d2/
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/auth/decorators.py", line 22, in _wrapped_view
    return view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 110, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/cache.py", line 43, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/util/views.py", line 41, in inner
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/views.py", line 352, in index
    return _index_bulk_op(request, course_key, chapter, section, position)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/views.py", line 528, in _index_bulk_op
    context['fragment'] = section_module.render(STUDENT_VIEW, section_render_context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/core.py", line 166, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1780, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1240, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/seq_module.py", line 189, in student_view
    rendered_child = child.render(STUDENT_VIEW, context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/core.py", line 166, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1780, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1240, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/vertical_block.py", line 46, in student_view
    rendered_child = child.render(STUDENT_VIEW, child_context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/core.py", line 166, in render
    return self.runtime.render(self, view, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1780, in render
    return self.__getattr__('render')(block, view_name, context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1240, in render
    return super(MetricsMixin, self).render(block, view_name, context=context)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 806, in render
    frag = view_fn(context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 532, in student_view
    return super(AdaptiveLibraryContentModule, self).student_view(context)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 337, in student_view
    for child in self._get_selected_child_blocks():
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 329, in _get_selected_child_blocks
    for block_type, block_id in self.selected_children():
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 497, in selected_children
    selected = self.get_selections_current_user(self.children)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/library_content_module.py", line 514, in get_selections_current_user
    pending_reviews = self.get_pending_reviews(self.anonymous_user_id)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/util/adaptive_learning.py", line 242, in get_pending_reviews
    pending_reviews_user = json.loads(response.content)
  File "/usr/lib/python2.7/json/__init__.py", line 326, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 384, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
[26/Dec/2016 07:34:08] "GET /courses/course-v1:OpenCraft+ACB1+OC-2150/courseware/519f9f58a85c4d2aba6182162b1c79e8/e94c284e0a104b77987a3a872367e9d2/ HTTP/1.1" 500 496062
```